packages/roosterjs-editor-dom/lib/edit/pasteTable.ts

```diff
@@ -0,0 +1,74 @@
+import findClosestElementAncestor from '../utils/findClosestElementAncestor';
+import getTagOfNode from '../utils/getTagOfNode';
+import toArray from '../jsUtils/toArray';
+import VTable from '../table/VTable';
+import { NodePosition, NodeType, TableOperation } from 'roosterjs-editor-types';
+
+/**
+ * @internal
+ * @param root The table to paste into
+ * @param nodeToInsert A Node containing the table to be inserted
+ * @param position The position to paste the table
+ */
+export default function pasteTable(
+    root: HTMLElement,
+    nodeToInsert: Node,
+    position: NodePosition,
+    range: Range
+) {
+    if (
+        (nodeToInsert.childNodes.length == 1 &&
+            getTagOfNode(nodeToInsert.childNodes[0]) == 'TABLE') ||
+        (nodeToInsert.childNodes.length == 2 &&
+            getTagOfNode(nodeToInsert.childNodes[0]) == 'TABLE' &&
+            getTagOfNode(nodeToInsert.childNodes[1]) == 'BR')
+    ) {
+        let rootNodeToInsert: Node | null = nodeToInsert;
+
+        if (rootNodeToInsert.nodeType == NodeType.DocumentFragment) {
+            let rootNodes = toArray(rootNodeToInsert.childNodes).filter(
+                (n: ChildNode) => getTagOfNode(n) != 'BR'
+            );
+            rootNodeToInsert = rootNodes.length == 1 ? rootNodes[0] : null;
+        }
+
+        let tdNode = findClosestElementAncestor(position.node, root, 'TD,TH');
+
+        // This is the table on the clipboard
+        let newTable = new VTable(<HTMLTableElement>rootNodeToInsert);
+        // This table is already on the editor
+        let currentCell = new VTable(<HTMLTableCellElement>tdNode);
+
+        let currentRow = currentCell.row!;
+        let currentCol = currentCell.col!;
+
+        let rows = currentRow + newTable.cells?.length!;
+        let columns = currentCol + newTable.cells?.[0].length!;
+
+        while (currentCell.cells!.length! < rows) {
+            currentCell.edit(TableOperation.InsertBelow);
+        }
+
+        while (currentCell.cells![0].length < columns) {
+            currentCell.edit(TableOperation.InsertRight);
+        }
+
+        for (let i = currentRow; i < rows; i++) {
+            for (let j = currentCol; j < columns; j++) {
+                let cell = currentCell.getCell(i, j);
+                let newcell = newTable.getTd(i - currentRow, j - currentCol);
+                if (!cell.spanAbove && !cell.spanLeft) {
+                    if (newcell!.hasChildNodes()) {
+                        // Prevent double <br>
+                        cell.td?.removeChild(cell.td.lastChild!);
+                        newcell!.childNodes.forEach(child => {
```

Comment:
I think an easier way is just replace the whole VCell object

Reply:
This creates formatting problems

Reply:
Can you elaborate more about the formatting problems?

Reply:
Currently the cell has the correct format it needs to have. By replacing the whole object I would be overwriting the format. The only thing that needs to change is the content.

Reply:
In that case, you can call moveChildNodes() here

Reply:
what do you mean by "checking twice"?

You just need to simply call moveChildNodes() to move anything we have from target TD to source TD, and replace anything we have there.

Reply:
I get a *Type 'null' is not assignable to type 'Node | undefined'.* warning if I do not add some check or add ! at the end of each node.

Reply:
Because the source TD can really be null. So that you must not add "!" otherwise there will be runtime exception.

We should only add "!" when we are pretty sure the object we are using MUST NOT be null and the compiler just not smart enough to detect that. Otherwise we should do null check.

And in this case, you need to create a TD element since TD doesn't exist in merged cells. And that is exactly the case strict mode can help us to prevent.

Reply:
What I am asking is moveChildNodes() makes sure that the nodes are not null and do have children, so even if they are it is already handled there.